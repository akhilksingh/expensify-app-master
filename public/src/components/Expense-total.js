import React from 'react';
import { connect } from 'react-redux';
import SelectExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpenseTotal = ({expenseCount,expenseTotal}) => {
    const formattedExpense = numeral(expenseTotal/100).format('$0,0.00');
    return (
        <div>
        {
            expenseCount === 0 ? (<p></p>) : 
            (
                <h2>
                <p> Viewing {expenseCount} Number of expense with total {formattedExpense} </p>
                </h2>
            )
        }
        </div>
    );
    

};
// //connect returns some High order component function which has access to Redux store
// // we pass what we want to connect in store , we have lot of information in store like state ,expenses, filters
const mapStateToStore = (state)=>
{
    const expenses = SelectExpenses(state.expenses,state.filters);
    const expenseTotal = expenses.reduce(
        (accumulator,currentValue)=>accumulator+currentValue.amount
        ,0
    );
    return {
        expenseCount : expenses.length,
        expenseTotal 
    };
};
export default connect(mapStateToStore)(ExpenseTotal);  //HOC

//const ConnectedExpenseList = connect()(ExpenseList);