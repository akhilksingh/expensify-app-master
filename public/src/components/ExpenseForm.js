import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/initialize'


// const now = moment();
// console.log(now.format(' Do, YYYY'));


export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description : props.expense ? props.expense.description : '',
            note : props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount/100).toString() : '',
            createdAt : props.expense ?  moment(props.expense.createdAt): moment(),
            calenderFocused : false,
            error :''
    
        };
    }

    onDescriptionChange=(e)=>{
        const description = e.target.value;
        this.setState(()=>({description}))
    };
    onNoteChange=(e)=> {
       // const note = e.target.value;
        //if you coment the above line , there will be errors as target value can't be directly used in callbacks, it should be saved somewhere first
        e.persist();
        this.setState(()=>({note :e.target.value}))
    };
    onChangeAmount=(e)=>{
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}))
        }

    }
    onDateChange =(createdAt)=>{
        this.setState(()=>({createdAt}))
    };
    onFocusChange=({focused})=>{
        this.setState(()=>({calenderFocused: focused}))
    };
    onSubmit = (e)=>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            //Set the error message to "Please provide valid messages" 
            this.setState(()=>({error : 'Please provide valid values for description and amount'}))
        }else{
            //clear the error 
            this.setState(()=>({error : ''}))
            //console.log("Submitted");
            this.props.onSubmit({
                description : this.state.description,
                amount : parseFloat(this.state.amount,10) * 100  ,//to convet decimal digits into cents , multiply by 100
                createdAt : this.state.createdAt.valueOf(),
                note : this.state.note
            })
        }
    }

    render(){
        return (
            <div>
              <form onSubmit={this.onSubmit}>
              {this.state.error && <p>{this.state.error}</p>}
              <input
               type="text"
               placeholder="Description"
               autoFocus
               value={this.state.description}
               onChange={this.onDescriptionChange}
               />
              
              <input 
              type="text"
              placeholder="Amount"
              value={this.state.amount}
              onChange={this.onChangeAmount}
              />
              <SingleDatePicker 
              date={this.state.createdAt}
              onDateChange={this.onDateChange}
              focused={this.state.calenderFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={()=>false}
              />

              <textarea
              placeholder="Add notes here for expense(optionl)"
              value = {this.state.note}
              onChange = {this.onNoteChange} >
              </textarea>
              <button >Add Expense</button>
              </form>
            </div>
        );

    }
}