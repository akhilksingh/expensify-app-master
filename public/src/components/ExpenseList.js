import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import SelectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
    
    {
        props.expenses.length === 0 ? (
            <p> No Expenses </p>
        ) : 
        (
                props.expenses.map((expense)=>{
                return <ExpenseListItem key={expense.id} {...expense} />
            })
            
        )
    }
    </div>

);
// //connect returns some High order component function which has access to Redux store
// // we pass what we want to connect in store , we have lot of information in store like state ,expenses, filters
const mapStateToStore = (state)=>
{
    return {
       expenses : SelectExpenses(state.expenses,state.filters)
    };
};
export default connect(mapStateToStore)(ExpenseList);  //HOC

const ConnectedExpenseList = connect()(ExpenseList);

//  export default ConnectedExpenseList;

 
// const App = (props) =>(
//     <div>
//     <p>Testing Connect Method</p>
//     {props.name}
//     </div>
// )
// const ConnectedExpenseList = connect((state)=>
// {
//     return {
//         name : 'Andrew'
//     }
// })(App);

// export default ConnectedExpenseList;