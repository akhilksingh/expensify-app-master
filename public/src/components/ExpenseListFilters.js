import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter} from '../actions/filters';
import {sortByDate,setStartDate,setEndDate} from '../actions/filters';
import {sortByAmount} from '../actions/filters';
import {DateRangePicker} from 'react-dates';


//Here we are going to implement the filters with Text input ,here we need to update the store , so guess what we need , STORE , to connect to store , we need connect() 
// connect returns a function which takes component as a parameter, connect takes a function as argument that determines what infomation of the store we want component to access , this function takes "store state" as first argument. 


//when you connect a component with redux , you get some other information as well i.e dispatch , here we have access to dispatch also which can dispatch actions to redux store
export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocussed : null
    };
    onDatesChange = ({startDate,endDate})=>{
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocussed)=>{
        this.setState(()=>({calendarFocussed}));
    };
    onTextChange = (e)=>{
        this.props.setTextFilter(e.target.value);
        //console.log(e.target.value);
    };
    onSortChange = (e)=>{
                
        if(e.target.value==='date'){
            this.props.sortByDate();
        }
        else if(e.target.value==='amount'){
            this.props.sortByAmount();
        }
    };

    render(){
        return (
            <div>
            <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
            <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                <option value="date">Date</option>
                <option value="amount">Amount</option>
            </select>
            <DateRangePicker
             startDate = {this.props.filters.startDate}
             endDate = {this.props.filters.endDate}
             onDatesChange = {this.onDatesChange}
             focusedInput = {this.state.calendarFocussed}
             onFocusChange = {this.onFocusChange}
             showClearDates={true}
             numberOfMonths={1}
             isOutsideRange={()=>false}

             />
            </div>
        );
    }
}


const mapStateToProps = (state) =>{
    return {
        filters : state.filters
    }

}
const mapDispathToProps = (dispatch)=>{
    return {
        setStartDate : (date)=>{dispatch(setStartDate(date))},
        setEndDate : (date)=>{dispatch(setEndDate(date))},
        setTextFilter: (text)=>{dispatch(setTextFilter(text))},
        sortByDate:()=>{dispatch(sortByDate())},
        sortByAmount:()=>{dispatch(sortByAmount())}
    };
}

export default connect(mapStateToProps,mapDispathToProps)(ExpenseListFilters);

