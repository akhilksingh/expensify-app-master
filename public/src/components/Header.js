import React from 'react';
import { BrowserRouter, Route ,Switch,Link,NavLink} from 'react-router-dom';
const Header = ()=>(
    <div>
    <h1>Expensify</h1>
    <h2>
    <NavLink to="/" activeClassName="is-active" exact="true">DashBoard </NavLink>
    <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
    <NavLink to="/help" activeClassName="is-active"> Help</NavLink>
    </h2>
    </div>
);

export default Header;