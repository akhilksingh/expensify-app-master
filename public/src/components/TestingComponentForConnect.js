import React from 'react';
import { connect } from 'react-redux';


const App = (props) =>(
    <div>
    <p>Testing Connect Method</p>
    {props.name}
    </div>
)
const ConnectedExpenseList = connect((state)=>
{
    return {
        name : 'Andrew'
    }
})(App);

export default ConnectedExpenseList;

