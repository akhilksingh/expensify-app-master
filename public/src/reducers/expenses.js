//we want to break this into multiple Reducers as this is complex object and we don't want to handle all request at one place , one handle expense and one handle filters


//Expenses Reducer
const expenseReducerDefaultState = [];

export default (state = expenseReducerDefaultState, action ) => {
    switch(action.type){
        case 'ADD_EXPENSE' :
        return [...state,action.expense];
        case 'REMOVE_EXPENSE' :
        return state.filter(({id})=>{ //filter doesn't change existing , if condition inside is true, then it will keep it , 
            return id !==action.id;
        });
        case 'EDIT_EXPENSE':
        return state.map((exp)=>{
            if(exp.id=== action.id)
            {
                return {
                    ...exp,
                    ...action.updates
                }
            }else
            {
                return exp;
            }
        })
        default:
        return state;
    }

};