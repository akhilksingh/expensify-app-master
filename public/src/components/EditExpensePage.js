import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {editExpense,removeExpense} from '../actions/expenses';




//EDITING THE EXPENSE with ID {props.match.params.id}
export class EditExpensePage extends React.Component {
    onSubmit = (expense)=>{
        //console.log('updated',expense);
        this.props.editExpense(this.props.expense.id,expense);
        this.props.history.push('/');
    }
    removeExpense = ()=>{
        this.props.removeExpense({id : this.props.expense.id});
        this.props.history.push('/');
    }
    render(){
        return (<div>
            <ExpenseForm 
            expense = {this.props.expense}
            onSubmit = {this.onSubmit}
            />
            <button 
            onClick={this.removeExpense}
            >Remove</button>
        </div>);
    }
    
}


const mapStateToProps=(state,props)=>{
    
    return {
        expense : state.expenses.find((expense)=> expense.id === props.match.params.id)
    };
};
const mapDispatchToProps = (dispatch,props)=>{
    return {
        editExpense : (id,expense)=>(dispatch(editExpense(id,expense))),
        removeExpense: (id)=>(dispatch(removeExpense({id : id.id})))
    };

}
export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage)