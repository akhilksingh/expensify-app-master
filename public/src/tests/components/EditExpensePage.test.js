import React from 'react';
import {shallow} from 'enzyme';
import {EditExpensePage} from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';
import moment from 'moment';

let editExpense , removeExpense, history ,wrapper,onSubmit;
beforeEach(()=>{
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = {push:jest.fn()};
    onSubmit = jest.fn();
    wrapper = shallow(<EditExpensePage 
        expense = {expenses[1]} 
        onSubmit={onSubmit}
        editExpense={editExpense}
        removeExpense={removeExpense}
        history ={history}
        />);
})
test('Should render Edit expense Page correctly',()=>{
    expect(wrapper).toMatchSnapshot();
});

test('Should handle remove expense',()=>{
    wrapper.find('button').simulate('click');
    expect(removeExpense).toHaveBeenLastCalledWith({id:expenses[1].id});
    expect(history.push).toHaveBeenLastCalledWith('/');
});

test('Should handle edit expense',()=>{
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id,expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
});

