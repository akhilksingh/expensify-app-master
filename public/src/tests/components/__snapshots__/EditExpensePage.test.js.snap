// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Edit expense Page correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 6,
        "createdAt": -345600000,
        "description": "Gas",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
